lib/instrument/transformer/nskeyed.js

```diff
@@ -0,0 +1,431 @@
+    val = new plistlib.UID(this.objects.length);
+    this.objects.push({
+      '$classes': [archiver],
+      '$classname': archiver
+    });
+    return val;
+  }
+
+  archive (obj) {
+    if (_.isUndefined(obj) || _.isNull(obj)) {
+      return NULL_UID;
+    }
+    const index = new plistlib.UID(this.objects.length);
+    if (PRIMITIVE_TYPES.includes(obj.constructor.name)) {
+      this.objects.push(obj);
+      return index;
+    }
+    const archiveObj = {};
+    this.objects.push(archiveObj);
+    this.encodeTopLevel(obj, archiveObj);
+    return index;
+  }
+
+  encode (val) {
+    if (NON_ENCODABLE_TYPES.includes(typeof val)) {
+      return val;
+    }
+    return this.archive(val);
+  }
+
+  encodeTopLevel (obj, archiveObj) {
+    if (obj instanceof Array) {
+      return this.encodeArray(obj, archiveObj);
+    } else if (obj instanceof Set) {
+      return this.encodeSet(obj, archiveObj);
+    } else if (obj instanceof Object) {
+      const objName = obj.constructor.name;
+      if (objName in UNARCHIVE_CLASS_MAP) {
+        archiveObj.$class = this.uidForArchiver(objName);
+        obj.encodeArchive(obj, new ArchivingObject(archiveObj, this));
+      } else {
+        return this.encodeDict(obj, archiveObj);
+      }
+    } else {
+      throw Error(`Unable to encode types: ${typeof obj}`);
+    }
+  }
+
+  encodeArray (objs, archiveObj) {
+    archiveObj.$class = this.uidForArchiver('NSArray');
+    archiveObj['NS.objects'] = objs.map(this.archive.bind(this));
+  }
+
+  encodeSet (objs, archiveObj) {
+    archiveObj.$class = this.uidForArchiver('NSSet');
+    archiveObj['NS.objects'] = objs.map(this.archive.bind(this));
+  }
+
+  encodeDict (obj, archiveObj) {
+    archiveObj.$class = this.uidForArchiver('NSDictionary');
+    archiveObj['NS.keys'] = _.keys(obj).map(this.archive.bind(this));
+    archiveObj['NS.objects'] = _.values(obj).map(this.archive.bind(this));
+  }
+
+  toBytes () {
+    if (this.objects.length === 1) {
+      this.archive(this.input);
+    }
+    const d = {
+      '$version': NSKeyedArchiveVersion,
+      '$archiver': NSKEYEDARCHIVER,
+      '$top': {'root': new plistlib.UID(1)},
+      '$objects': this.objects
+    };
+    return bplistCreate(d);
+  }
+}
+
+/**
+ * Creates NSKeyed Buffer from an object
+ * @param {Object} inputObject
+ * @returns {Buffer} NSKeyed Buffer
+ */
+function archive (inputObject) {
+  return new Archive(inputObject).toBytes();
+}
+
+/**
+ * Parses NSKeyed Buffer into JS Object
+ * @param {Buffer} inputBytes NSKeyed Buffer
+ * @returns {Object} JS Object
+ */
+function unarchive (inputBytes) {
+  return new Unarchive(inputBytes).toObject();
+}
+
+/**
+ * Update unknown NSKeyedArchive types for packing/unpacking
+ * @param {String} name packing/unpacking key name
+ * @param {BaseArchiveHandler} subClass inherit from BaseArchiveHandler class
+ */
+function updateNSKeyedArchiveClass (name, subClass) {
+  if (!(Object.getPrototypeOf(subClass) === BaseArchiveHandler)) {
```

Comment:
In languages with dynamic typing I would prefer to rather use duck-style check to explicit type checks. For example, this condition could check if the given class/instance have encodeArchive/decodeArchive methods

Reply:
Like this?
```
if (!_.isFunction(subClass.prototype.decodeArchive) && !_.isFunction(subClass.prototype.encodeArchive))
```

Reply:
yes. In strictly typed languages one would use an interface for such purprose

Reply:
you could also use `subClass.prototype?.decodeArchive` to make the check safer